File: leetcode-cn/0167.0_Two_Sum_II_-_Input_Array_Is_Sorted.py
'''
two pointers
T: O(N)
S: O(1)

执行用时：48 ms, 在所有 Python3 提交中击败了45.40% 的用户
内存消耗：15.7 MB, 在所有 Python3 提交中击败了49.99% 的用户
通过测试用例：21 / 21
'''
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        i, j = 0, len(numbers) - 1
        while i < j:
            s = numbers[i] + numbers[j]
            if s == target:
                return [i + 1, j + 1]
            elif s < target:
                i += 1
            else:
                j -= 1
        return []

'''
执行用时：48 ms, 在所有 Python3 提交中击败了73.60% 的用户
内存消耗：17.2 MB, 在所有 Python3 提交中击败了15.85% 的用户
通过测试用例：23 / 23
'''
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        n = len(numbers)
        i, j = 0, n - 1
        while i < j:
            _sum = numbers[i] + numbers[j]
            if _sum == target:
                return [i + 1, j + 1]
            elif _sum < target:
                i += 1
            else:
                j -= 1
            
            

File: java/0167.0_Two_Sum_II_-_Input_Array_Is_Sorted.java
/*
two pointers
T: O(N)
S: O(1)

执行用时：1 ms, 在所有 Java 提交中击败了92.70% 的用户
内存消耗：44.3 MB, 在所有 Java 提交中击败了24.58% 的用户
通过测试用例：21 / 21
*/
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int[] indices = new int[2];
        int i = 0, j = numbers.length - 1;
        while (i < j) {
            int s = numbers[i] + numbers[j];
            if (s == target) {
                indices[0] = i + 1;
                indices[1] = j + 1;
                return indices;
            } else if (s > target) {
                j--;
            } else {
                i++;
            }
        }
        return indices;
    }
}



/*
执行用时：1 ms, 在所有 Java 提交中击败了92.70% 的用户
内存消耗：44.5 MB, 在所有 Java 提交中击败了5.10% 的用户
通过测试用例：21 / 21
*/
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        for (int i = 0, j = numbers.length - 1; i < j; ) {
            int sum = numbers[i] + numbers[j];
            if (sum == target) {
                return new int[] {i + 1, j + 1};
            } else if (sum > target) {
                j--;
            } else {
                i++;
            }
        }
        return null;
    }
}

